fdbserver/BlobManager.actor.cpp

```diff
@@ -1312,6 +1317,68 @@ ACTOR Future<Void> blobWorkerRecruiter(
 	}
 }
 
+// TODO: make this its own role if it becomes a performance bottleneck for BM
+ACTOR Future<Void> blobPruner(BlobManagerData* bmData) {
+	state Version lastDeleteVersion = 0;
+	loop {
+		Version deleteVersion = waitNext(bmData->deleteRequests.getFuture());
+
+		// if this is an old prune request, then do nothing since we
+		// already saw a newer prune request
+		if (lastDeleteVersion >= deleteVersion) {
+			continue;
+		}
+
+		lastDeleteVersion = deleteVersion;
+
+		// TODO: implement deletion logic
+	}
+}
+
+// When a new blob manager is created, we want to prune right away so that
+// if the previous blob manager died right before pruning, we don't wait approximately 2R
+// versions before pruning
+ACTOR Future<Void> blobPrunerScheduler(BlobManagerData* bmData) {
+	loop {
+		printf("prune scheduler\n");
+		state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(bmData->db);
+
+		Version currVersion = wait(tr->getReadVersion());
+		Version newPruneVersion = std::max<Version>(0, currVersion - SERVER_KNOBS->BG_FILE_RETENTION_PERIOD);
+		if (newPruneVersion <= bmData->latestPruneVersion) {
+			delay(SERVER_KNOBS->BG_FILE_PRUNER_CADENCE);
+			continue;
+		}
+		printf("newPruneVersion: %lld, latestPruneVersion: %lld\n", newPruneVersion, bmData->latestPruneVersion);
+
+		ASSERT(newPruneVersion > bmData->latestPruneVersion);
+
+		// although it might seem a little early to set this, we need to so that new
+		// blob workers which are created while reqs are in fligh use the latest prune
+		bmData->latestPruneVersion = newPruneVersion;
+
+		// broadcast new prune version to all workers
+		std::vector<Future<Void>> acks;
+		for (auto& [id, bwi] : bmData->workersById) {
+			PruneFilesRequest req(bmData->epoch, bmData->latestPruneVersion);
+			acks.emplace_back(bwi.pruneFiles.getReply(req)); // TODO: use some sort of priority here?
+		}
+
+		// wait until all blob workers have agreed to stop reading older than the new prune version
+		// TODO: do we have to do anything here to take care of a BW that died before it sent ACK
```

Comment:
I was unsure if `waitForAll` handles this. I was looking around the codebase and I couldn't find an example of us doing something special to handle this

Reply:
Will add error handling accordingly